File: servoParser.js
const fs = require("fs");
const traceJson = JSON.parse(
  fs.readFileSync(
    "C:/Users/William/code/school/servo/layout_trace-1.json",
    "utf-8"
  )
);

function parseJson(json) {
  const root = traceJson.post;

  const recurse = element => {
    const {
      size: { block: heightAus, inline: widthAus },
      start: { b: yAus, i: xAus }
    } = element.data.base.position;
    const values = {
      x: xAus / 60,
      y: yAus / 60,
      width: widthAus / 60,
      height: heightAus / 60
    };
    values.children = Array.from(element.data.base.children).map(child =>
      recurse(child)
    );
    return values;
  };

  return recurse(root.children[0]);
}

const parsed = parseJson(traceJson);

fs.writeFileSync(
  "C:/Users/William/code/school/servo/layout_trace-1-parsed.json",
  JSON.stringify(parsed)
);

File: servo_parser.py
def parse_servo_json(servo_json):
    root = servo_json

    def recurse(element, parent_x, parent_x_margin, parent_y):
        data = element['data']['block_flow'] \
            if element['class'] == 'Flex' \
            else element['data']
        position = data['base']['position']
        border_box = data['fragment']['border_box']
        x = (position['start']['i'] / 60) + (border_box['start']['i'] /
                                             60) + parent_x - parent_x_margin
        y = (position['start']['b'] / 60) + (border_box['start']['b'] /
                                             60) + parent_y
        margin_x = (data['fragment']['margin']['inline_start'] / 60)
        values = {
            'x': x,
            'y': y,
            'width': border_box['size']['inline'] / 60,
            'height': border_box['size']['block'] / 60
        }

        values['children'] = list(
            map(lambda child: recurse(child, x, margin_x, y), data['base']['children']))

        return values

    return recurse(root['children'][0], 0, 0, 0)
